### PR TITLE
Added Feature: Prefix decoding for wav2vec2 models

### LINKED ISSUE #11283: Beam search decoding and language model integration for Wav2Vec2ForCTC models
1. AFAIK, `Wav2Vec2ForCTCTokenizer.decode` method only provides greedy decoding. Is there a Beamsearch implementation for CTC available yet? 
2. Also, as it is a common norm in ASR modelling, language models are also generally added on top of the acoustic model. It would also be nice to have a possibility of appending a pretrained Language model which gets taken into consideration at the beamsearch decoding time. Not sure if there's an out-of-box solution implemented for that yet?

I'm also aware of efforts to integrate a language model in #10794 and have had a look at the notebook [here](https://github.com/voidful/huggingface_notebook/blob/main/xlsr_gpt.ipynb). Although it is a nice, simple way to integrate an LM, it is suboptimal when considering CTC semantics. A more appropriate approach would be the one described in [this](https://arxiv.org/pdf/1408.2873.pdf) paper and explained in [this](https://distill.pub/2017/ctc/) distilpub blog. Would be great to have these features added (if they are already not there and I somehow missed them).

### PR DESCRIPTION
**Pull request by deepang17**
# What does this PR do?
Added the code for prefix decoding for wav2vec2 based models.

Fixes #11283 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @patil-suraj

### CODE REVIEW

**Comment by deepang17** [comment]
1. Currently the code support prefix decoding without LM. I am still working to integrate the kenlm version.
Problem faced currently: While running the kenlm based decoder it throws SegmentationFault which needs to be fixed. Also, I created a custom kenlm and tried to run the code, but it stops without throwing any error at line `results = self.decoder.decode(emissions_ptr, T, N)` I am currently trying to fix.

2. Shall I create a .sh or .txt file to guide on how to install flashlight dependencies?

**Comment by deepang17** [comment]
Performance: `model: facebook/wav2vec2-base-100h`, `dataset: timit_asr clean`

- Greedy : `wer: 0.164`
- Prefix (viterbi): `wer: 0.163`
- Prefix with lm (kenlm): `Work in progress`

**Comment by deepang17** [comment]
Performance: 
                      Model:- `facebook/wav2vec2-base-960h`
                      Dataset:- `timit_asr`, `clean`, `test[:5%]`
                      Viterbi Decoding:- `wer: 0.115`
                      KenLM Decoding:- `wer: 0.098`

**Comment by patrickvonplaten** [comment]
Wuhuhu! This is an amazing contribution @deepang17 - Super exciting to merge this notebook :-) And yes, it would be great if you could add a section to the README.md that explains how to use your script + maybe with some results (using Prefix decoding vs. not using it on *e.g.* Librispeech evaluation). I'm also very happy to help you here, if you need anything

**Review by patrickvonplaten — APPROVED** [review]
Longterm we could even think about merging this into `src/transformers/models/wav2vec2/` - but for now this is great!

**Comment by deepang17** [comment]
Thank you for the appreciation. I Will do the required changes to `README.md` and push a commit soon.

**Comment by samuelazran** [comment]
> * [x]   Currently the code supports prefix decoding without LM. I am still working to integrate the kenlm version.
> 
> Problem faced currently: I created a custom kenlm and tried to run the code, but it stops without throwing any error at line `results = self.decoder.decode(emissions_ptr, T, N)` I am currently trying to fix it. (RESOLVED)
> 
> * [x]  Shall I create a .sh or .txt file to guide on how to install flashlight dependencies?
@deepang17
Did you pushed that fix? I've tried your code and it is crushing at the "self.decoder.decode". What was your fix?

What is the status of this PR?

**Comment by deepang17** [comment]
> > * [x]    Currently the code supports prefix decoding without LM. I am still working to integrate the kenlm version.
> > 
> > Problem faced currently: I created a custom kenlm and tried to run the code, but it stops without throwing any error at line `results = self.decoder.decode(emissions_ptr, T, N)` I am currently trying to fix it. (RESOLVED)
> > 
> > * [x]   Shall I create a .sh or .txt file to guide on how to install flashlight dependencies?
> 
> @deepang17
> Did you pushed that fix? I've tried your code and it is crushing at the "self.decoder.decode". What was your fix?
> 
> What is the status of this PR?

You can fix it by replacing `!cmake .. -DCMAKE_BUILD_TYPE=Release -DKENLM_MAX_ORDER=20 -DCMAKE_POSITION_INDEPENDENT_CODE=ON` to `!cmake ..`

**Comment by bayethiernodiop** [comment]
> test[:5%]

Hi @deepang17  thanks for this. I have an issue, could you have any guidance please.
I mangage to make this work for the timit dataset as you did, but when I try to adapt it to another language WOLOF in this case, I have garbage output when using either viterbi or my kenlm model. but I have correct output with greedy decoding.
output with greedy : 'lu tax ngay sarax njàccaar yi bàyyi silmaxa yi' => good
output with viterbi : 'kçóñhfólàhyóxh hfóltjnnhh óyiórjyyióxikchfhóyió' => bad

Any help please.

**Comment by samuelazran** [comment]
> > > * [x]     Currently the code supports prefix decoding without LM. I am still working to integrate the kenlm version.
> > > 
> > > Problem faced currently: I created a custom kenlm and tried to run the code, but it stops without throwing any error at line `results = self.decoder.decode(emissions_ptr, T, N)` I am currently trying to fix it. (RESOLVED)
> > > 
> > > * [x]    Shall I create a .sh or .txt file to guide on how to install flashlight dependencies?
> > 
> > 
> > @deepang17
> > Did you pushed that fix? I've tried your code and it is crushing at the "self.decoder.decode". What was your fix?
> > What is the status of this PR?
> 
> You can fix it by replacing `!cmake .. -DCMAKE_BUILD_TYPE=Release -DKENLM_MAX_ORDER=20 -DCMAKE_POSITION_INDEPENDENT_CODE=ON` to `!cmake ..`

Can you please publish a Google Colab or a bash script to do the installation? I could't figure out where to do the change you suggested in the build, I'v used the Google Colab example from flashlight.

File: examples/research_projects/wav2vec2/run_wav2vec2_eval_with_lm.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            ).input_values.to("cuda")
+        else:
+            input_values = processor(
+                batch["speech"], sampling_rate=batch["sampling_rate"], return_tensors="pt"
+            ).input_values
+
+        with torch.no_grad():
+            logits = model(input_values).logits
+
+        target_dictionary = [t for t in processor.tokenizer.get_vocab().keys()]
```
Comment (joaoalvarenga) [inline examples/research_projects/wav2vec2/run_wav2vec2_eval_with_lm.py:260]: Testing W2lViterbiDecoder i figure out that this list must by ordered by original token index.
```suggestion
        vocab = processor.tokenizer.get_vocab()
        target_dictionary = sorted(vocab, key=lambda k: vocab[k])
```

**Comment by tommy19970714** [comment]
@deepang17 Thank you for your amazing work!
I made Google Colab to reproduce this pull request. @samuelazran You can check this.
https://colab.research.google.com/drive/1HHEBS3I4biQ8ZDyfJDtHi4E4onOtYe46?usp=sharing

Viterbi Decoding works well, But KenLM Decoding has the following error.

```
  File "run_wav2vec2_eval_with_lm.py", line 292, in <module>
    main()
  File "run_wav2vec2_eval_with_lm.py", line 281, in main
    results = selected_dataset.map(map_to_result)
  File "/usr/local/lib/python3.7/dist-packages/datasets/arrow_dataset.py", line 1606, in map
    desc=desc,
  File "/usr/local/lib/python3.7/dist-packages/datasets/arrow_dataset.py", line 176, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/datasets/fingerprint.py", line 397, in wrapper
    out = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/datasets/arrow_dataset.py", line 1911, in _map_single
    example = apply_function_on_filtered_inputs(example, i, offset=offset)
  File "/usr/local/lib/python3.7/dist-packages/datasets/arrow_dataset.py", line 1826, in apply_function_on_filtered_inputs
    function(*fn_args, effective_indices, **fn_kwargs) if with_indices else function(*fn_args, **fn_kwargs)
  File "run_wav2vec2_eval_with_lm.py", line 265, in map_to_result
    decoder = W2lKenLMDecoder(eval_args, target_dictionary)
  File "run_wav2vec2_eval_with_lm.py", line 201, in __init__
    self.lm = KenLM(args.kenlm_model, self.word_dict)
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. flashlight.lib.text.flashlight_lib_text_decoder.KenLM(path: str, usr_token_dict: fl::lib::text::Dictionary)

Invoked with: None, <flashlight.lib.text.flashlight_lib_text_dictionary.Dictionary object at 0x7fe0ef7294b0>
```

@deepang17  Do you know this error?  It exactly gives as an argument of flashlight.lib.text.flashlight_lib_text_decoder.KenLM the dict obtained from flashlight.lib.text.dictionary.create_word_dict.

**Comment by patrickvonplaten** [comment]
@deepang17 - do you have updates regarding the README.md script? :-) I can take over the PR by next week otherwise!

**Comment by deepang17** [comment]
Hello @patrickvonplaten, Sorry for the delay. I was occupied due to some personal issues. I am on the verge of completing the README.md. I will commit the updated README soon.

**Comment by tommy19970714** [comment]
@deepang17  Any updates?

**Comment by shiva1393** [comment]
@deepang17 Any updates?

**Comment by patrickvonplaten** [comment]
This PR seems to be stuck since quite some time now. Is anyone interested in finishing / testing this PR? 

Might be better to start fresh otherwise with a blog post ) colab that explains how to make a complete ASR end-to-end system - cc @anton-l

**Comment by hbasafa** [comment]
@patrickvonplaten 
I'm in! I've searched this topic but it seems there is no official implementation on this topic, and It would be so nice to add this feature. If this feature is still in the backlog, I would be happy to contribute. Looking forward to your alert!

**Comment by hbasafa** [comment]
Nice one! I will check it out.

As I was in hurry, I've already used [this code](https://github.com/hbasafa/py-ctc-decode) that could be easily installed vie pip.
The code sample is also provided in [here](https://github.com/hbasafa/wav2vec_decode)
And Now I also focused on other decoding strategies to add there.

However, Thank you for sharing!

**Comment by machakos23** [comment]
> Hey @hbasafa,
> 
> I'm now working on this topic full time.
> 
> We will most likely foster a closer collaboration between [pyctcdecode](https://github.com/kensho-technologies/pyctcdecode) and Transformers. [Here](https://github.com/patrickvonplaten/Wav2Vec2_PyCTCDecode) is a github repo that shows how to use `pyctcdecode` with Wav2Vec2 for LM supported decoding. It works quite well with KenLM.

hi @patrickvonplaten  - this is great news. Where is the best place to follow your progress?

**Comment by patrickvonplaten** [comment]
This PR: https://github.com/huggingface/transformers/pull/14339

If everything goes well, we should have a merged version to master by the end of this week

**Pull request closed without merging**